tests/nightly/straight_dope/straight_dope_test_utils.py

```diff
@@ -0,0 +1,106 @@
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+#pylint: disable=no-member, too-many-locals, too-many-branches, no-self-use, broad-except, lost-exception, too-many-nested-blocks, too-few-public-methods, invalid-name
+"""
+    This file tests and ensures that all straight dope notebooks run
+    without warning or exception.
+
+    env variable MXNET_TEST_KERNEL controls which kernel to use when running
+    the notebook. e.g: `export MXNET_TEST_KERNEL=python2`
+"""
+import io
+import os
+import re
+import shutil
+import subprocess
+import sys
+
+#TODO(vishaalk): Find a cleaner way to import this notebook.
+sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', '..', 'utils'))
+from notebook_test import run_notebook
+
+EPOCHS_REGEX = r'epochs\s+=\s+[0-9]+'  # Regular expression that matches 'epochs = #'
+GIT_PATH = '/usr/bin/git'
+GIT_REPO = 'https://github.com/zackchase/mxnet-the-straight-dope'
+KERNEL = os.getenv('MXNET_TEST_KERNEL', None)
+NOTEBOOKS_DIR = os.path.join(os.path.dirname(__file__), 'tmp_notebook')
+
+def _test_notebook(notebook, override_epochs=True):
+    """Run Jupyter notebook to catch any execution error.
+
+    Args:
+        notebook : string
+            notebook name in folder/notebook format
+        epochs : boolean
+            whether or not to override the number of epochs to 1
+    Returns:
+        True if the notebook runs without warning or error.
+    """
+    if override_epochs:
+        _override_epochs(notebook)
+    return run_notebook(notebook, NOTEBOOKS_DIR, kernel=KERNEL, temp_dir=NOTEBOOKS_DIR)
+
+
+def _override_epochs(notebook):
+    """Overrides the number of epochs in the notebook to 1 epoch. Note this operation is idempotent.
+
+    Args:
+        notebook : string
+            notebook name in folder/notebook format
+
+    """
+    notebook_path = os.path.join(*([NOTEBOOKS_DIR] + notebook.split('/'))) + ".ipynb"
+
+    # Read the notebook and set epochs to num_epochs
+    with io.open(notebook_path, 'r', encoding='utf-8') as f:
+        notebook = f.read()
+
+    # Set number of epochs to 1
+    modified_notebook = re.sub(EPOCHS_REGEX, 'epochs = 1', notebook)
+
+    # Replace the original notebook with the modified one.
+    with io.open(notebook_path, 'w', encoding='utf-8') as f:
+        f.write(modified_notebook)
+
+
+def _download_straight_dope_notebooks():
+    """Downloads the Straight Dope Notebooks.
+
+    Returns:
+        True if it succeeds in downloading the notebooks without error.
+    """
+    print('Cleaning and setting up notebooks directory "{}"'.format(NOTEBOOKS_DIR))
+    shutil.rmtree(NOTEBOOKS_DIR, ignore_errors=True)
+
+    cmd = [GIT_PATH,
+           'clone',
+           GIT_REPO,
+           NOTEBOOKS_DIR]
+
+    proc = subprocess.Popen(
+        cmd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.STDOUT)
+    (out, _) = proc.communicate()
```

Comment:
Nice! Is there a way we could add a timeout here? Git tends to not be very nice to us in terms of reliability..

Reply:
Added a commit to add a timeout.